### PR TITLE
LVGL All The Things!

### PR DESCRIPTION
**Pull request by mutatrum**
Combined PR for the LVGL implementation. For the bare bones implementation of LVGL alone, please have a look at #434 first. This PR pushes it a lot further, by leveraging LVGL functionalities and cleanup up a lot of code. Some noticeable:

 * Long text scrolling labels;
 * Screen transition animations;
 * Button click and hold handling;
 * System task and user input task are gone;

Implementation wise the biggest change is that a lot of data is passed through GLOBAL_STATE, which cleans up a lot of display logic form other places.

Things to do/check:

 * ~~IP address changes;~~ :white_check_mark: 
 * ~~Pool changes, both fallback as well as configuration changes;~~ :white_check_mark: 
 * ~~Block found screen;~~ :white_check_mark: 
 * ~~Invalid ASIC model. When does this happen and what should be displayed?~~ :white_check_mark: 
 * ~~Checking of overheat mode reset is moved to power management task;~~ :white_check_mark: 
 * ~~Show connection screen again if AP mode is enabled;~~ :white_check_mark: 

image

### CODE REVIEW

**Comment by mutatrum** [comment]
Found the issue. I2C error on panel init stops it from initialising LVGL, which is needed later on.

File: main/lv_conf.h [hunk 1]
```diff
@@ -0,0 +1,1127 @@
...
+
+/* clang-format off */
+#if 1 /*Set it to "1" to enable content*/
+
+#ifndef LV_CONF_H
+#define LV_CONF_H
+
+/*If you need to include anything here, do it inside the `__ASSEMBLY__` guard */
+#if  0 && defined(__ASSEMBLY__)
+#include "my_include.h"
```
Comment (WantClue) [inline main/lv_conf.h:22]: I think it's safe to comment this out

File: main/screen.c [hunk 2]
```diff
@@ -0,0 +1,312 @@
...
+static lv_obj_t * create_scr_configure(SystemModule * module) {
+    lv_obj_t * scr = lv_obj_create(NULL);
+
+    lv_obj_set_flex_flow(scr, LV_FLEX_FLOW_COLUMN);
+    lv_obj_set_flex_align(scr, LV_FLEX_ALIGN_SPACE_EVENLY, LV_FLEX_ALIGN_START, LV_FLEX_ALIGN_START);
+
+    lv_obj_t *label1 = lv_label_create(scr);
+    lv_obj_set_width(label1, LV_HOR_RES);
+    lv_label_set_long_mode(label1, LV_LABEL_LONG_SCROLL_CIRCULAR);
+    lv_label_set_text(label1, "Welcome to your new BitAxe! Connect to the configuration Wifi and connect the BitAxe to your network.");
```
Comment (WantClue) [inline main/screen.c:92]: remove camel cases

File: main/lv_conf.h (hunk 1 above)
Comment (mutatrum) [reply main/lv_conf.h:22]: This whole file is a copy of the template supplied by lvgl, with only the changes to defiate from the defaults which minimises the changes. The other option would be to make a tiny lv_conf.h file, which I can take a stab at.

**Comment by mutatrum** [comment]
> Upgraded my Supra without a screen and it unfortunately didn't come back online. I'll plug in the USB later to get the logs.

I got the button working without the screen. It does complain a lot on boot:

```
I (1293) display: Install panel IO
I (1293) display: Install SSD1306 panel driver
E (1293) i2c.master: I2C hardware NACK detected
E (1293) i2c.master: I2C transaction unexpected nack detected
E (1303) i2c.master: s_i2c_synchronous_transaction(917): I2C transaction failed
E (1303) i2c.master: i2c_master_multi_buffer_transmit(1166): I2C transaction failed
E (1313) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(193): i2c transaction failed
E (1323) lcd_panel.ssd1306: panel_ssd1306_init(151): io tx param SSD1306_CMD_SET_MULTIPLEX failed
E (1333) display: Panel init failed, no display connected?
I (1333) display: Initialize LVGL
I (1343) LVGL: Starting LVGL task
E (1343) i2c.master: I2C hardware NACK detected
E (1343) i2c.master: I2C transaction unexpected nack detected
E (1353) i2c.master: s_i2c_synchronous_transaction(917): I2C transaction failed
E (1353) i2c.master: i2c_master_multi_buffer_transmit(1166): I2C transaction failed
E (1363) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(193): i2c transaction failed
E (1373) lcd_panel.ssd1306: panel_ssd1306_mirror(237): io tx param SSD1306_CMD_MIRROR_X_ON/OFF failed
E (1383) i2c.master: I2C hardware NACK detected
E (1383) i2c.master: I2C transaction unexpected nack detected
E (1393) i2c.master: s_i2c_synchronous_transaction(917): I2C transaction failed
E (1393) i2c.master: i2c_master_multi_buffer_transmit(1166): I2C transaction failed
E (1403) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(193): i2c transaction failed
E (1413) lcd_panel.ssd1306: panel_ssd1306_draw_bitmap(196): io tx param SSD1306_CMD_SET_COLUMN_RANGE failed
W (1423) display: No display found.
W (1423) SystemModule: OLED init failed!
I (1433) input: Install button driver
I (1433) gpio: GPIO[0]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:3 
```

File: main/lv_conf.h (hunk 1 above)
Comment (mutatrum) [reply main/lv_conf.h]: Made a minimal `lv_conf.h`.

**Comment by eandersson** [comment]
> > Upgraded my Supra without a screen and it unfortunately didn't come back online. I'll plug in the USB later to get the logs.
> 
> I got the button working without the screen. It does complain a lot on boot:
> 
> ```
> I (1293) display: Install panel IO
> I (1293) display: Install SSD1306 panel driver
> E (1293) i2c.master: I2C hardware NACK detected
> E (1293) i2c.master: I2C transaction unexpected nack detected
> E (1303) i2c.master: s_i2c_synchronous_transaction(917): I2C transaction failed
> E (1303) i2c.master: i2c_master_multi_buffer_transmit(1166): I2C transaction failed
> E (1313) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(193): i2c transaction failed
> E (1323) lcd_panel.ssd1306: panel_ssd1306_init(151): io tx param SSD1306_CMD_SET_MULTIPLEX failed
> E (1333) display: Panel init failed, no display connected?
> I (1333) display: Initialize LVGL
> I (1343) LVGL: Starting LVGL task
> E (1343) i2c.master: I2C hardware NACK detected
> E (1343) i2c.master: I2C transaction unexpected nack detected
> E (1353) i2c.master: s_i2c_synchronous_transaction(917): I2C transaction failed
> E (1353) i2c.master: i2c_master_multi_buffer_transmit(1166): I2C transaction failed
> E (1363) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(193): i2c transaction failed
> E (1373) lcd_panel.ssd1306: panel_ssd1306_mirror(237): io tx param SSD1306_CMD_MIRROR_X_ON/OFF failed
> E (1383) i2c.master: I2C hardware NACK detected
> E (1383) i2c.master: I2C transaction unexpected nack detected
> E (1393) i2c.master: s_i2c_synchronous_transaction(917): I2C transaction failed
> E (1393) i2c.master: i2c_master_multi_buffer_transmit(1166): I2C transaction failed
> E (1403) lcd_panel.io.i2c: panel_io_i2c_tx_buffer(193): i2c transaction failed
> E (1413) lcd_panel.ssd1306: panel_ssd1306_draw_bitmap(196): io tx param SSD1306_CMD_SET_COLUMN_RANGE failed
> W (1423) display: No display found.
> W (1423) SystemModule: OLED init failed!
> I (1433) input: Install button driver
> I (1433) gpio: GPIO[0]| InputEn: 1| OutputEn: 0| OpenDrain: 0| Pullup: 1| Pulldown: 0| Intr:3 
> ```
> 
> This also means self test succeeds even without a display. It does throw the same errors.

Complaining is fine, as long as it does not spam (or even worse impact performance) after boot is done.

**Comment by mutatrum** [comment]
> Complaining is fine, as long as it does not spam (or even worse impact performance) after boot is done.

No, it only complains at boot. All screen handling and updating is done in an `lv_timer`, which is not started if the display is not attached. 

https://github.com/skot/ESP-Miner/blob/78ec3c8202631b44d18a527b1e99497d18303b04/main/screen.c#L306-L317

**Review by WantClue — APPROVED** [review]

File: main/display.c [hunk 3]
```diff
@@ -0,0 +1,133 @@
...
+#include "global_state.h"
+#include "nvs_config.h"
+#include "i2c_bitaxe.h"
+#include "driver/i2c_master.h"
+#include "driver/i2c_types.h"
+#include "esp_lcd_panel_ssd1306.h"
+
+#define SSD1306_I2C_ADDRESS    0x3C
+
+#define LCD_H_RES              128
```
Comment (WantClue) [inline main/display.c:22]: as you suggested in dc already, we might need to figure out a good way for having different display sizes

**Review by WantClue — APPROVED** [review]

**Review by eandersson — APPROVED** [review]

File: main/display.c (hunk 3 above)
Comment (mutatrum) [reply main/display.c:22]: See #572.

**Pull request merged**